ex8/k_sawada/my_functions/hmm.py

```diff
@@ -0,0 +1,82 @@
+import numpy as np
+
+
+class HMM:
+    def __init__(self, states, initial_state_probability,
+                 transition_probability, output_probability):
+        """HMM
+
+        Args:
+            states (int):
+                number of states
+            initial_state_probability (ndarray, axis=(state)):
+                initial state probability array
+            transition_probability (ndarray, axis=(before, after)):
+                transition probability matrix
+            output_probability (ndarray, axix=(state, output label)):
+                output probability matrix
+        """
+        self.states = states
+        self.initial_state_probability = initial_state_probability
+        self.transition_probability = transition_probability
+        self.output_probability = output_probability
+    
+    def forward(self, outputs):
+        """likelihoods based on Forward algorithm
+        likelihoods is equal to joint probability
+
+        Args:
+            outputs (ndarray, axis=(time)):
+                outputs series
+
+        Returns:
+            float:
+                likelihoods
+        """
+        length = len(outputs)
+        alpha = np.zeros((length, self.states))
+        
+        # base stage
+        alpha[0] = self.initial_state_probability * \
+            self.output_probability[:, outputs[0]]
+        
+        # recursion stage
+        for t in range(1, length):
+            alpha[t] = (alpha[t - 1].reshape((1, self.states)) @
+                        self.transition_probability).flatten() * \
+                self.output_probability[:, outputs[t]]
+        return np.sum(alpha[-1])
+    
+    def viterbi(self, outputs):
+        """likelihoods based on Viterbi algorithm
+        likelihoods is equal to joint probability
+
+        Args:
+            outputs (ndarray, axis=(time)):
+                outputs series
+
+        Returns:
+            float:
+                likelihoods
+        """
+        length = len(outputs)
+        psi_probability = np.zeros((length, self.states))
+        psi_states = np.zeros((length, self.states))
+
+        # base stage
+        psi_probability[0] = self.initial_state_probability * \
+            self.output_probability[:, outputs[0]]
+        # psi_states[0] is already zeros
+
+        # recursion stage
+        for t in range(1, length):
+            psi_probability[t] = \
+                np.max(
+                    self.transition_probability *
+                    psi_probability[t - 1].reshape(self.states, 1), axis=0) * \
```

Comment:
こういう書き方もできます。軸の追加をするなら具体的なshape記述が必要ないこちらのほうが便利なのでおすすめです
```suggestion
                    psi_probability[t - 1, :, np.newaxis], axis=0) * \
```

Reply:
https://qiita.com/rtok/items/10f803a226892a760d75 https://note.nkmk.me/python-numpy-newaxis/
を読んだ感じ, reshapeでサイズ1を追加する操作をnp.newaxisで置き換えられる…と理解しました.  
なので `psi_probability[t - 1, :, np.newaxis], axis=0)` でなくて, `psi_probability[t - 1, np.newaxis], axis=0)` のほうがピッタリなのかなと思ったのですが, どうなんでしょうか? (どちらも問題なく動きました.)

Reply:
`psi_probability[t - 1, :, np.newaxis]`のshapeは`(self.states,1)`
`psi_probability[t - 1, np.newaxis]　   ` のshapeは`(1,self.states)`
です。なので、両者は `self.transition_probability` に掛けた場合の結果が異なると思います。（どちらの場合もブロードキャスト自体はできるのでエラーは出ません）

Reply:
ほんとですね… 修正したら正解率が上がり, 正しそうな結果になりました